### DISCUSSION TITLE
About default feature range in Kinematic Observation

### DISCUSSION THREAD

**Discussion by mhtb32** [eleurent/highway-env#169] [Ideas]
Hi!

I'm curious where the [default](https://github.com/eleurent/highway-env/blob/b94c2d058e9ff6a1764d7e16b72bc0822fd1fe15/highway_env/envs/common/observation.py#L147) features range in `KinematicObservation` come from? To the point I know, the `x` value can be from 0 to `maximum_duration * MAX_SPEED`, but in that line, it is considered in (-5 * MAX_SPEED, 5 * MAX_SPEED) range. So I can't understand:
- Why is the negative range also considered for `x` and `y`?
- Why is maximum duration not considered?
- Why five times of MAX_SPEED?

Again I don't understand why the speed range is two times the MAX_SPEED.

Thank you

**Comment by mhtb32** [comment]
And one other thing: The `PERCEPTION_DISTANCE` in the `AbstractEnv` is considered six times the MAX_SPEED, which compared to the values above is a vast quantity and is almost like global knowledge of the environment.

**Comment by eleurent** [comment]
1. in general roads can have any direction, so it should be handled. But even on a horizontal road, you may want to perceive vehicles all around you, not just in front (x) and to your right (?) (y)
2. and 3. well, maximum duration is chosen to 5s (hardcoded, yes). So a unit distance is the maximum the ego-vehicle can travel in 5s.
4. the speed range is 2*max speed, because its the tightest upper bound on v1 - v2, when |v1|,|v2| <= max_speed. So its mainly for relative observations.

**Comment by eleurent** [comment]
> The PERCEPTION_DISTANCE in the AbstractEnv is considered six times the MAX_SPEED, which compared to the values above is a vast quantity

Well, it's 6 max_speed rather than 5 max_speed, not that vast a difference. But I agree that it' an inconsistency.

> almost like global knowledge of the environment.

If you want to plan on a horizon of 5-6s, you probably need to see that far. 100-150m seems reasonable to me, in the case of highway driving.

**Comment by mhtb32** [comment]
> 1. in general roads can have any direction, so it should be handled. But even on a horizontal road, you may want to perceive vehicles all around you, not just in front (x) and to your right (?) (y)

So if I understood right, the observations are relative tho the vehicle's position and are not absolute. My question was with the pre assumption that observations are absolute.

**Comment by eleurent** [comment]
Well it can be both, depending on whether the field `absolute` is set to true or false.
But even in the case of absolute coordinates, in some environments the frame origin is located such that vehicles can have negative coordinates. The horizontal roads of highway-v0 are a special case.

**Comment by mhtb32** [comment]
> But even in the case of absolute coordinates, in some environments the frame origin is located such that vehicles can have negative coordinates. The horizontal roads of highway-v0 are a special case.

Could you mention a code line in which you've changed the origin? I wanted to see how it is possible.

**Comment by mhtb32** [comment]
> 2\. and 3. well, the maximum duration is chosen to 5s (hardcoded, yes). So a unit distance is the maximum the ego-vehicle can travel in 5s.

I want to reward my agent based on normalized distance from a predefined goal, but isn't the point of normalization that we can always expect the normalized value won't exceed 1? Let's, in this case, say that the vehicle is at `x=300`m absolute position, then its normalized `x` is 2.

So can I conclude that in such cases, we should limit the duration to 5 seconds to make sure the agent's observation remains normalized? (Or of course, specify a courser feature range)

**Comment by eleurent** [comment]
> Could you mention a code line in which you've changed the origin? I wanted to see how it is possible.

You cannot really change the origin, but rather the position of the road network (relative to the origin). In intersection-v0 or roundabout-v0 for instance, the origin is in the middle of the intersection/roundbout.

> but isn't the point of normalization that we can always expect the normalized value won't exceed 1? 

Well, not necessarily. It depends on why you are normalizing, really. If it is related to model conditioning for function approximation, it doesn't have to be a hard constraint, and people often normalize by dividing by the standard deviation (which means you can still get a value >= 2 with ~13% probability).
See e.g. https://www.jeremyjordan.me/batch-normalization/, https://towardsdatascience.com/why-data-should-be-normalized-before-training-a-neural-network-c626b7f66c7d

I think it is up to you to choose the proper feature range / scale depending on your particular needs.

**Comment by mhtb32** [comment]
Thank you very much for the fruitful discussion
